source/websocket.c

```diff
@@ -0,0 +1,395 @@
+    int error_code,
+    void *user_data);
+
+/* When WebSocket._binding is GC'd, release the native websocket pointer.
+ * It will close (if necessary) on its way to the grave */
+static void s_websocket_capsule_destructor(PyObject *capsule) {
+    struct aws_websocket *websocket = PyCapsule_GetPointer(capsule, s_websocket_capsule_name);
+    aws_websocket_release(websocket);
+}
+
+/* Kick off websocket connection.
+ * WebSocket._binding does not get returned from this function,
+ * it gets delivered later via the _WebSocketCore._on_connection_setup() callback */
+PyObject *aws_py_websocket_client_connect(PyObject *self, PyObject *args) {
+    (void)self;
+
+    struct aws_byte_cursor host;    /* s# */
+    uint16_t port;                  /* H */
+    PyObject *handshake_request_py; /* O */
+    PyObject *bootstrap_py;         /* O */
+    PyObject *socket_options_py;    /* O */
+    PyObject *tls_options_py;       /* O */
+    PyObject *proxy_options_py;     /* O */
+    int enable_read_backpressure;   /* p - boolean predicate */
+    Py_ssize_t initial_read_window; /* n */
+    PyObject *websocket_core_py;    /* O */
+
+    if (!PyArg_ParseTuple(
+            args,
+            "s#HOOOOOpnO",
+            &host.ptr,
+            &host.len,
+            &port,
+            &handshake_request_py,
+            &bootstrap_py,
+            &socket_options_py,
+            &tls_options_py,
+            &proxy_options_py,
+            &enable_read_backpressure,
+            &initial_read_window,
+            &websocket_core_py)) {
+        return NULL;
+    }
+
+    /* First, wrangle args that don't require any cleanup if things go wrong... */
+
+    /* required bootstrap */
+    struct aws_client_bootstrap *bootstrap = aws_py_get_client_bootstrap(bootstrap_py);
+    if (bootstrap == NULL) {
+        return NULL;
+    }
+
+    /* required socket_options */
+    struct aws_socket_options socket_options;
+    if (aws_py_socket_options_init(&socket_options, socket_options_py) == false) {
+        return NULL;
+    }
+
+    /* optional tls_options */
+    struct aws_tls_connection_options *tls_options = NULL;
+    if (tls_options_py != Py_None) {
+        tls_options = aws_py_get_tls_connection_options(tls_options_py);
+        if (tls_options == NULL) {
+            return NULL;
+        }
+    }
+
+    /* optional proxy_options */
+    bool has_proxy_options = proxy_options_py != Py_None;
+    struct aws_http_proxy_options proxy_options;
+    if (has_proxy_options) {
+        if (aws_py_http_proxy_options_init(&proxy_options, proxy_options_py) == false) {
+            return NULL;
+        }
+    }
+
+    /* required handshake_request */
+    struct aws_http_message *handshake_request = aws_py_get_http_message(handshake_request_py);
+    if (handshake_request == NULL) {
+        return NULL;
+    }
+
+    /* required window size */
+    if (initial_read_window < 0) {
+        PyErr_Format(PyExc_ValueError, "'initial_read_window' cannot be negative");
```

Comment:
trivial: should we move this check to python? Feel like it will be more clear to the user.